src/validops.cu

```diff
@@ -0,0 +1,59 @@
+#include <gdf/gdf.h>
+#include <gdf/utils.h>
+#include <thrust/device_ptr.h>
+#include <thrust/transform_reduce.h>
+#include <thrust/functional.h>
+
+gdf_error gdf_count_nonzero_mask(gdf_valid_type * masks, int num_rows, int * count)
+{
+  // Number of masks in the array of bitmasks
+  const size_t num_masks = gdf_get_num_chars_bitmask(num_rows);
+
+  // Device lambda to count the number of valid bits in each mask
+  auto mask_bit_counter = [masks, num_masks, num_rows] __device__ (int index)
+  {
+
+    constexpr int BITS_PER_MASK = sizeof(gdf_valid_type) * 8;
+
+    gdf_valid_type count = 0;
+    gdf_valid_type mask = masks[index];
+    if(index < num_masks - 1)
+    {
+      // FIXME: If gdf_valid_type was 32 bits, we could use __popc
```

Comment:
Can't you process in groups of 4 bytes (except the last one) and use __popc?

Reply:
Indeed, see updated implementation.